### DISCUSSION TITLE
CsWinRT 2.0 and .NET 5

### DISCUSSION THREAD

**Discussion by manodasanW** [microsoft/CsWinRT#1232] [Announcements]
With [.NET 5’s end of support on ](https://devblogs.microsoft.com/dotnet/dotnet-5-end-of-support-update/)May 10, 2022  , the Windows SDK projection brought in by the .NET 5 SDK is no longer updated. .NET 6 and later SDKs continue to be updated. With CsWinRT bringing in new features such as IL trimming, the generated projections are starting to use APIs that are only available in or can only be guaranteed to be available in applications targeting .NET 6 and later. Due to this, CsWinRT plans to drop support for .NET 5 and move to CsWinRT 2.0. CsWinRT 1.6.5 will be the last update which supports .NET 5 and any customers who still need .NET 5 compatible projections may continue to use it.

**Comment by michael-hawker** [comment]
FYI @azchohfi @Sergio0694

**Comment by Sergio0694** [comment]
Does the 2.0.0 version indicate (as per semantic versioning) that it will have the liberty of making some breaking changes and correct some unfortunate design decisions of the first release, and improve the API shape in general? Just to make a few random examples: 

Having `ConcurrentDictionary<TKey, TValue>` (and 2 of them) in the public API surface of `IWinRTObject`:

https://github.com/microsoft/CsWinRT/blob/a59ba547021dd740240ac7cd6ab5ac982f5c65a9/src/WinRT.Runtime/IWinRTObject.net5.cs#L153

https://github.com/microsoft/CsWinRT/blob/a59ba547021dd740240ac7cd6ab5ac982f5c65a9/src/WinRT.Runtime/IWinRTObject.net5.cs#L169

Function pointers in `IUnknownVtbl` from using a managed signature and causing a marshalling stub to be generated:

https://github.com/microsoft/CsWinRT/blob/a59ba547021dd740240ac7cd6ab5ac982f5c65a9/src/WinRT.Runtime/Interop/IUnknownVftbl.cs#L17-L22

The `IObjectReference` type exposing a protected bool field:

https://github.com/microsoft/CsWinRT/blob/a59ba547021dd740240ac7cd6ab5ac982f5c65a9/src/WinRT.Runtime/ObjectReference.cs#L25

Or is 2.0.0 just to indicate the fact .NET 5 is being dropped? ...I hope it's the former? 😄

**Reply by dotMorten** [reply]
The real question is how would this affect downstream? Ie WinAppSDK takes a dependency here and also gets a breaking change, in addition to all 3rd party class libraries that takes a dependency on that. I’d thread VERY lightly with introducing any breaking changes beyond the tfm change - the ecosystem is still trying to catch up to WinAppSDK and I honestly don’t think they can afford to break people.

**Reply by manodasanW** [reply]
@Sergio0694 For 2.0.0, we don't have any plans to do any breaking changes other than dropping .NET 5 support for the reasons @dotMorten mentioned.

@Sergio0694, regarding your `IUnknownVtbl` point, can you expand on it?  Is this due to the use of ref and out in the signature or due to something else and what is the impact there with respect to marshaling stubs?

**Reply by Sergio0694** [reply]
> "For 2.0.0, we don't have any plans to do any breaking changes other than dropping .NET 5"

I'm a bit sad to hear that. There's several issues in CsWinRT that if we can't address with a major release I don't know when we'll ever be able to. It'd be a bit of a shame to just remain forever stuck with those without being able to fix them 🥲

Also, I don't think dropping an old TFM (especially if it's EOL, but even if that wasn't the case) is a breaking change at all according to semantics versioning. Users with a compatible TFM will see neither a source nor a binary breaking change. Users in an old TFM just won't be able to update, but won't be broken either. If they bump the TFM, they they would also not see any breaking changes whatsoever. Honestly just dropping .NET 5 might have been done in a minor release as well. 

> "Is this due to the use of ref and out in the signature"

Correct. I was chatting about this with @aaronrobinsonmsft a while back too: basically if your function pointer signature is not blittable (which is the case here, due to those GC refs), the JIT will be forced to emit a marshalling stub at the invocation site (in this case, it will pin those refs for you). This is obviously bad for performance and not something you'd want such a fundamental building block to have as overhead. It's also not particularly AOT friendly for the same reason (though it does also work on NativeAOT, of course).

I know that CsWinRT as a whole, at least for now, doesn't support disabling built-in marshalling yet, but that's one of the things that would fall apart if a user tried to enable that. Especially with the focus recently towards trimming, partial reflection mode etc., it would be nice if we could gradually get to a place where a fully trimmed, no built-in marshalling, NativeAOT CsWinRT app could be done 🙂

**Reply by dotMorten** [reply]
> I'm a bit sad to hear that.

But I'm happy to hear that. As a 3rd party vendor I can't keep playing catchup to breaking changes - especially not to a platform that doesn't have a huge customer base currently - Something like that could cause us to drop support if we find it too unstable and cost-prohibitive to continue to support.
I'd rather that instead of defaulting to "lets break everyone", they took at look at providing an alternative to the things that are blocking you.

Perhaps some of the new NET7 stuff that can do better blitting and remove the need for stub generation in the interop will address that specific issue?

**Reply by dotMorten** [reply]
> Honestly just dropping .NET 5 might have been done in a minor release as well.

I think that's a fair point. After all that's what WinAppSDK is doing too.

**Reply by manodasanW** [reply]
@Sergio0694 I looked at where we make use of the QueryInterface function pointer from IUnknownVftbl and all of it uses seems to be within WinRT.Runtime and is not used directly in any of our generated projections.  Due to that, I believe we should be able to address that concern without breaking any existing projections or consumers of those projections.

**Reply by sylveon** [reply]
> The real question is how would this affect downstream? Ie WinAppSDK takes a dependency here and also gets a breaking change, in addition to all 3rd party class libraries that takes a dependency on that. I’d thread VERY lightly with introducing any breaking changes beyond the tfm change - the ecosystem is still trying to catch up to WinAppSDK and I honestly don’t think they can afford to break people.

WASDK has happily broken C++ users in minor releases before, so the ecosystem is already dealing with breaking changes. What's one more breaking change to the list?

**Reply by hez2010** [reply]
@manodasanW
Actually we consider WASDK 1.2 with CsWinRT 2.0 to be the first usable version of WinUI 3 apps, because it's the first version that aligns WinUI 2.8 styles and supports IL trimming, and also, the first version that doesn't support net5.0. 
I think once we miss the chance of making breaking changes in CsWinRT 2.0, it will be much harder to do so in the future.

**Comment by jlaanstra** [comment]
Thoughts on getting rid of the As* methods as part of 2.0 and do the mapping automatically like we already do for IIterable->IEnumerable, IVector ->IList, Uri->Uri etc? Few examples that come to mind:

- Automatically mapping Task<T> and Task to IAsyncOperation/IAsyncAction without the need for AsAsyncOperation()/AsAsyncAction()?
- Autmatically mapping stream types without the need for AsStream()/AsRandomAccessStream()?

**Reply by manodasanW** [reply]
It is an interesting idea to do a custom type mapping for Task and Streams (it is a bit more involved than the others), but for 2.0 we are avoiding changes that can break compat with existing projections or their consumers.
